contracts/oracles/yieldspace/PoolOracle.sol

```diff
@@ -0,0 +1,146 @@
+
+    // mapping from pool address to a list of ratio observations of that pool
+    mapping(address => Observation[]) public poolObservations;
+
+    constructor(
+        uint256 windowSize_,
+        uint256 granularity_,
+        uint256 minTimeElapsed_
+    ) {
+        require(granularity_ > 1, "GRANULARITY");
+        require((periodSize = windowSize_ / granularity_) * granularity_ == windowSize_, "WINDOW_NOT_EVENLY_DIVISIBLE");
+        windowSize = windowSize_;
+        granularity = granularity_;
+        minTimeElapsed = minTimeElapsed_;
+    }
+
+    /// @dev calculates the index of the observation corresponding to the given timestamp
+    /// @param timestamp The timestamp to calculate the index for
+    /// @return index The index corresponding to the `timestamp`
+    function observationIndexOf(uint256 timestamp) public view returns (uint256 index) {
+        uint256 epochPeriod = timestamp / periodSize;
+        return epochPeriod % granularity;
+    }
+
+    /// @dev returns the observation from the oldest epoch (at the beginning of the window) relative to the current time
+    /// @param pool Address of pool for which the observation is required
+    /// @return The oldest observation available for `pool`
+    function getOldestObservationInWindow(address pool) public view returns (Observation memory) {
+        if (poolObservations[pool].length == 0) {
+            revert NoObservationsForPool(pool);
+        }
+
+        unchecked {
+            uint256 observationIndex = observationIndexOf(block.timestamp);
+            uint256 length = poolObservations[pool].length;
+            uint256 oldestObservationIndex;
+            // can't possible overflow
+            for (uint256 i; i < length; ++i) {
+                // no overflow issue. if observationIndex + 1 overflows, result is still zero.
+                oldestObservationIndex = (++observationIndex) % granularity;
+                // If no data exists for this index, check the next one,
+                // this should only happen during the first timeWindow for a given pool
+                // If the elapsedTime is bigger than the timeWindow,
+                // we also check the next one in case only this particular observation is stale
+                if (block.timestamp - poolObservations[pool][oldestObservationIndex].timestamp < windowSize) {
+                    return poolObservations[pool][oldestObservationIndex];
+                }
+            }
+
+            revert MissingHistoricalObservation(pool);
+        }
+    }
+
+    // @inheritdoc IPoolOracle
+    function update(address pool) public override {
+        // populate the array with empty observations (oldest call only)
+        for (uint256 i = poolObservations[pool].length; i < granularity; i++) {
+            poolObservations[pool].push();
+        }
+
+        // get the observation for the current period
+        uint256 index = observationIndexOf(block.timestamp);
```

Comment:
imagine we're in a 1 hour periodSize.  what happens if update hasn't been called for 3 hours? 🤔